File: test_parse.py
import re

with open('ReportHistory-28315610.html', 'r', encoding='utf-16le') as f:
    content = f.read()

# find all trs
trs = re.findall(r'<tr[^>]*>(.*?)</tr>', content, re.DOTALL | re.IGNORECASE)

in_positions = False
trades = []

for tr in trs:
    th = re.search(r'<th[^>]*>.*?<b>(.*?)</b>', tr, re.DOTALL | re.IGNORECASE)
    if th:
        text = th.group(1).strip()
        if text in ['Positions', 'Closed Transactions']:
            in_positions = True
            continue
        elif text in ['Orders', 'Deals', 'Open Positions']:
            in_positions = False
            continue

    if in_positions:
        # get all tds
        tds = re.findall(r'<td([^>]*)>(.*?)</td>', tr, re.DOTALL | re.IGNORECASE)
        # filter out hidden
        tds = [td for td in tds if 'class="hidden"' not in td[0]]
        
        if len(tds) >= 13:
            time_open = re.sub(r'<[^>]*>', '', tds[0][1]).strip()
            if re.match(r'^\d{4}\.\d{2}\.\d{2} \d{2}:\d{2}:\d{2}$', time_open):
                order_id = re.sub(r'<[^>]*>', '', tds[1][1]).strip()
                time_close = re.sub(r'<[^>]*>', '', tds[8][1]).strip()
                profit = re.sub(r'<[^>]*>', '', tds[12][1]).strip()
                commission = re.sub(r'<[^>]*>', '', tds[10][1]).strip()
                swap = re.sub(r'<[^>]*>', '', tds[11][1]).strip()
                
                if order_id and time_close and time_close != 'cancelled':
                    trades.append({
                        'profit': float(profit.replace(',', '') or 0),
                        'commission': float(commission.replace(',', '') or 0),
                        'swap': float(swap.replace(',', '') or 0)
                    })

total = sum([t['profit'] + t['commission'] + t['swap'] for t in trades])
print(f"Total PnL from Positions: {total}")
print(f"Number of trades: {len(trades)}")

File: test_parse.js
const fs = require('fs');
const { JSDOM } = require('jsdom');

const htmlStr = fs.readFileSync('ReportHistory-28315610.html', 'utf-16le');
const dom = new JSDOM(htmlStr);
const doc = dom.window.document;
const rows = doc.querySelectorAll('tr');
const trades = [];

let inPositionsSection = false;

rows.forEach(row => {
    const cells = Array.from(row.querySelectorAll('td')).filter(c => !c.classList.contains('hidden'));
    
    const thB = row.querySelector('th b')?.textContent;
    if (thB === 'Positions' || thB === 'Closed Transactions') {
        inPositionsSection = true;
        return;
    }
    if (thB === 'Orders' || thB === 'Deals' || thB === 'Open Positions') {
        inPositionsSection = false;
        return;
    }

    if (inPositionsSection && cells.length >= 13) {
        const timeOpenStr = cells[0]?.textContent?.trim();
        if (timeOpenStr && /^\d{4}\.\d{2}\.\d{2} \d{2}:\d{2}:\d{2}$/.test(timeOpenStr)) {
            const orderId = cells[1]?.textContent?.trim();
            const symbol = cells[2]?.textContent?.trim();
            const type = cells[3]?.textContent?.trim();
            const volume = cells[4]?.textContent?.trim();
            const priceOpen = cells[5]?.textContent?.trim();
            const sl = cells[6]?.textContent?.trim();
            const tp = cells[7]?.textContent?.trim();
            const timeCloseStr = cells[8]?.textContent?.trim();
            const priceClose = cells[9]?.textContent?.trim();
            const commission = cells[10]?.textContent?.trim();
            const swap = cells[11]?.textContent?.trim();
            const profit = cells[12]?.textContent?.trim();

            if (orderId && timeCloseStr && timeCloseStr !== 'cancelled') {
                trades.push({
                    orderId,
                    profit: parseFloat(profit?.replace(/,/g, '') || '0'),
                    commission: parseFloat(commission?.replace(/,/g, '') || '0'),
                    swap: parseFloat(swap?.replace(/,/g, '') || '0')
                });
            }
        }
    }
});

let totalPnl = 0;
trades.forEach(t => {
    totalPnl += t.profit + t.commission + t.swap;
});
console.log('Parsed trades count:', trades.length);
console.log('Total PnL:', totalPnl);
